homeassistant/components/light/tplink.py

```diff
@@ -138,25 +138,31 @@ def update(self):
                 self.smartbulb.state == self.smartbulb.BULB_STATE_ON)
             if self._name is None:
                 self._name = self.smartbulb.alias
-            if self.smartbulb.is_dimmable:
+            if self._supported_features & SUPPORT_BRIGHTNESS:
                 self._brightness = brightness_from_percentage(
                     self.smartbulb.brightness)
-            if self.smartbulb.is_variable_color_temp:
+            if self._supported_features & SUPPORT_COLOR_TEMP:
                 if (self.smartbulb.color_temp is not None and
                         self.smartbulb.color_temp != 0):
                     self._color_temp = kelvin_to_mired(
                         self.smartbulb.color_temp)
-            if self.smartbulb.is_color:
+            if self._supported_features & SUPPORT_RGB_COLOR:
                 self._rgb = hsv_to_rgb(self.smartbulb.hsv)
             if self.smartbulb.has_emeter:
                 self._emeter_params[ATTR_CURRENT_CONSUMPTION] \
                     = "%.1f W" % self.smartbulb.current_consumption()
                 daily_statistics = self.smartbulb.get_emeter_daily()
                 monthly_statistics = self.smartbulb.get_emeter_monthly()
-                self._emeter_params[ATTR_DAILY_CONSUMPTION] \
-                    = "%.2f kW" % daily_statistics[int(time.strftime("%d"))]
-                self._emeter_params[ATTR_MONTHLY_CONSUMPTION] \
-                    = "%.2f kW" % monthly_statistics[int(time.strftime("%m"))]
+                try:
+                    self._emeter_params[ATTR_DAILY_CONSUMPTION] \
+                        = "%.2f kW" % daily_statistics[int(
+                            time.strftime("%d"))]
+                    self._emeter_params[ATTR_MONTHLY_CONSUMPTION] \
+                        = "%.2f kW" % monthly_statistics[int(
+                            time.strftime("%m"))]
+                except KeyError:
```

Comment:
I'm not sure if there could be a case where only either or is available, but in my opinion this will do fine for now. The real issue should be fixed in upstream by returning 0 when no information is available.